WIP Widgets version2, compilation tests

GillesDuvert opened the pull request:
only test this time

GillesDuvert committed:
Merge pull request #81 from gnudatalanguage/master

align master on true one

GillesDuvert committed:
Merge pull request #82 from gnudatalanguage/master

align

GillesDuvert committed:
initial import, new version Widgets using wxWidgets

alaingdl commented:
Wouhaou, great

But sorry, could you repeat the simplest way (serie of CLI instructions) to download that, for example starting from zero
or from an already cloned master ? (same situation in fact)

thanks 

A.

Ps: I may also be able to test on one OSX where I already compiled GDL within Widgets

GillesDuvert committed:
oops, this must be made simpler, for the moment, lets use the old complitacte multi-device scheme

GillesDuvert committed:
ok, needs checkvalid!

GillesDuvert committed:
no use to patch winstream to mimic [w]xstream behaviour, this device will be removed in favor of wxwidgets

GillesDuvert committed:
force wx when wxwidgets is present.

GillesDuvert committed:
better graphics device initialisation?

GillesDuvert committed:
removed windows-only ifdef that apparently is not supported contrary to documentation.

GillesDuvert committed:
try using wxAutoBufferedPaintDC for plots/draw 1) seems to make it 100 times faster than X11 and 2) perhaps refresh will be ok on windows? keep fingers crossed.

GillesDuvert commented:
retry

GillesDuvert commented:
excellent, with this scheme aborting the tests will still produce the artifacts.

GillesDuvert committed:
solved a few remaining problems. Enabled plots wider than the screen, they get scrollbars.
Probably not sufficient for correct use under Windows.

GillesDuvert committed:
test with paintDC see if pleases windows

GillesDuvert committed:
remove old test_widgets; try wxClientDC for windows to avoid wxFAIL_MSG( wxT("wxPaintDCImpl may be created only in EVT_PAINT handler!") );

GillesDuvert committed:
try fixing Windows problems

GillesDuvert commented:
![image](https://user-images.githubusercontent.com/33936193/107132168-fd7f3c00-68dc-11eb-9454-079f25140c8c.png)

GillesDuvert committed:
Be X always

GillesDuvert committed:
try forcing setFont on all widgets

GillesDuvert committed:
test without UpdateGui() . corrects bug destruction of tabbed bases.

GillesDuvert committed:
assure some fontsize if cannot retrieve it from system?

GillesDuvert commented:
Works OK on Windows.

pjb7687 commented:
Works well!

GillesDuvert committed:
corrected how DEVICE,get_fontname behaves (should not give NULL) and wxW implementation now takes into account the pattern set by SET_FONT.

GillesDuvert committed:
added option --widget-free when strict size compatibility with IDL is not needed (better looks).
test message for the problem of fonts under windows

GillesDuvert committed:
force test windows

GillesDuvert committed:
again

GillesDuvert committed:
again, test

GillesDuvert committed:
simplified fontSize treatment, see how it works on Windows.

GillesDuvert committed:
shoudl return nativefontinfoUSERdesc.

GillesDuvert committed:
Added widget_info,string_size=[...,...] and other string-size related stuff

GillesDuvert committed:
patches tentatively adressing windows-specific problems

GillesDuvert committed:
simplify, complexify...

GillesDuvert committed:
oops this shoudl be done differently --- I'll leave it to somebody with a windows OS.

GillesDuvert commented:
![image](https://user-images.githubusercontent.com/33936193/107558552-379d5600-6bdb-11eb-8839-c674d98ab23c.png)

pjb7687 commented on src/gdlwidget.hpp:
```diff
@@ -39,9 +39,6 @@
 #include <wx/notebook.h>
 #include <wx/dcbuffer.h>
-#if __WXMSW__
-#include <wx/msw/wx.rc>  //should provide manifest, but is linker OK?
```
Does it require manifest? AFAIK you can use `windres`...

GillesDuvert replied:
@pjb7687 the widgets have no Look & Feel on Windows, because somewhere in the build process, one shoudl provide a gdl.exe.manifest for the windows executable.
But I cannot find an example of how to do that, wxWidget documentation seems to believe everyone adds manifests just like they are breathing and do not need any explanation or example.
At least, #include <wx/msw/wx.rc> only was a failure.
Probably you can help?

pjb7687 replied:
Added 2 commits below... It should work now.

GillesDuvert replied:
thanks a lot, the look and feel has improved!

pjb7687 committed:
add wx.rc to sources

pjb7687 committed:
Merge branch 'master' into WidgetsVersion2

pjb7687 committed:
improve compilation release builds

pjb7687 commented:
I've recently fixed the macOS build, but this patch seems incompatible on macOS. Here you can see some test failures...
https://travis-ci.com/github/gnudatalanguage/gdl/jobs/482830812

maynardGK commented:
I have been attempting to incorporate the new GDL source as well as update my MSYS package:  I could run the CI-built distros in the previous BASH, but because of windows' upgrades or something I lost the ability to properly connect with the libraries I held to run, i.e.  the Q32 version of GraphicsMagick.

So that's where I am at the moment, and I notice again about the BSD/XDR package,
the basic distro provides an xdr,h which is different from the xdr.h given by the hdf4
package:
````
$ pacman -Qo /mingw64/include/xdr.h
/mingw64/include/xdr.h is owned by mingw-w64-x86_64-hdf4 4.2.15-2
$ cat /mingw64/include/xdr.h | head
/*      $NetBSD: xdr.h,v 1.19 2000/07/17 05:00:45 matt Exp $    */

/*
 * Copyright (c) 2009, Sun Microsystems, Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * - Redistributions of source code must retain the above copyright notice,
 *   this list of conditions and the following disclaimer.

greg@i7Wx MSYS /d/sources/bsd-xdr-1.0.0
$ cat rpc/xdr.h | head

/*
 * Copyright (c) 2009, Sun Microsystems, Inc.
 * All rights reserved
 ````
````
41,45c41,42
< #ifndef __XDR_HEADER__
< #define __XDR_HEADER__
<
< #include "H4api_adpt.h"
<
---
> #ifndef _RPC_XDR_H
> #define _RPC_XDR_H
46a44,48
> #include <rpc/types.h>
>>
> #ifdef __cplusplus
> extern "C" {
> #endif
57,60c59,60
<  *    bool_t
<  *    xdrproc(xdrs, argresp)
<  *        XDR *xdrs;
<  *        <type> *argresp;
---
>  *        bool_t
>  *        xdrproc(XDR *xdrs, <type> *argresp)
````

GillesDuvert commented:
@maynardGK actually, installing the last artifact of this build on a windows 10, i cannot open images at all,  **so libmagick is not working**. I can open tiff 8 bit images however, this does not use graphicsmagick/imagemagick. But no 24-bit tiff images (the tests in testsuite/tiff), but this is true also for the linux version, the error is:
`% Parser syntax error: unexpected token: READ_TIFF`
meaning a NULL has been returned by whatever function is called when reading 24 bit Tiff images.

pjb7687 commented:
> @maynardGK actually, installing the last artifact of this build on a windows 10, i cannot open images at all, **so libmagick is not working**. I can open tiff 8 bit images however, this does not use graphicsmagick/imagemagick. But no 24-bit tiff images (the tests in testsuite/tiff), but this is true also for the linux version, the error is:
> `% Parser syntax error: unexpected token: READ_TIFF`
> meaning a NULL has been returned by whatever function is called when reading 24 bit Tiff images.

This should be fixed now: https://github.com/gnudatalanguage/gdl/issues/883
If you still can't open any images please open the issue again.

pjb7687 commented:
> I have been attempting to incorporate the new GDL source as well as update my MSYS package: I could run the CI-built distros in the previous BASH, but because of windows' upgrades or something I lost the ability to properly connect with the libraries I held to run, i.e. the Q32 version of GraphicsMagick.
> 
> So that's where I am at the moment, and I notice again about the BSD/XDR package,
> the basic distro provides an xdr,h which is different from the xdr.h given by the hdf4
> package:
> 
> ```
> $ pacman -Qo /mingw64/include/xdr.h
> /mingw64/include/xdr.h is owned by mingw-w64-x86_64-hdf4 4.2.15-2
> $ cat /mingw64/include/xdr.h | head
> /*      $NetBSD: xdr.h,v 1.19 2000/07/17 05:00:45 matt Exp $    */
> 
> /*
>  * Copyright (c) 2009, Sun Microsystems, Inc.
>  * All rights reserved.
>  *
>  * Redistribution and use in source and binary forms, with or without
>  * modification, are permitted provided that the following conditions are met:
>  * - Redistributions of source code must retain the above copyright notice,
>  *   this list of conditions and the following disclaimer.
> 
> greg@i7Wx MSYS /d/sources/bsd-xdr-1.0.0
> $ cat rpc/xdr.h | head
> 
> /*
>  * Copyright (c) 2009, Sun Microsystems, Inc.
>  * All rights reserved
> ```
> 
> ```
> 41,45c41,42
> < #ifndef __XDR_HEADER__
> < #define __XDR_HEADER__
> <
> < #include "H4api_adpt.h"
> <
> ---
> > #ifndef _RPC_XDR_H
> > #define _RPC_XDR_H
> 46a44,48
> > #include <rpc/types.h>
> >>
> > #ifdef __cplusplus
> > extern "C" {
> > #endif
> 57,60c59,60
> <  *    bool_t
> <  *    xdrproc(xdrs, argresp)
> <  *        XDR *xdrs;
> <  *        <type> *argresp;
> ---
> >  *        bool_t
> >  *        xdrproc(XDR *xdrs, <type> *argresp)
> ```

I don't think the difference in the XDR is a big problem. HDF4 just has its own implementation of XDR, which is dependent on other internal HDF4 headers, and this can be considered as something independent from the normal XDR library that we use (i.e. BSD-XDR). Also, the header location is different (one is in `include`, the other one is in `include/rpc`) and they do not overwrite each other.

Regarding GraphicsMagick, then I think it would be nice to build a Q24 or Q32 version in the CI script. But I have a question: Isn't Q16~Q24~ enough for most cases?

pjb7687 commented:
> I've recently fixed the macOS build, but this patch seems incompatible on macOS. Here you can see some test failures...
> https://travis-ci.com/github/gnudatalanguage/gdl/jobs/482830812

Interesting, I've tried compiling it on macOS Catalina, it worked very well. 100% test passed.
I uploaded a compiled binary so that you can test it: https://filebin.net/aakbf0y84sq1ybr6/gdl-full-debug-catalina.zip?t=voqy7v6y

FYI the binary was compiled on Github Actions: 
https://github.com/actions/virtual-environments/blob/main/images/macos/macos-10.15-Readme.md

maynardGK commented:
> Regarding GraphicsMagick, then I think it would be nice to build a Q24 or Q32 version in the CI script. But I have a question: Isn't Q24 enough for most cases?

The goal I had was for a Q16 version, since, I beliieve, that is the level most distros have.  The MSYS2/mingw64 version was Q8 last time I checked ... if that were changed to Q16 the world would be a better place.

pjb7687 commented:
> > Regarding GraphicsMagick, then I think it would be nice to build a Q24 or Q32 version in the CI script. But I have a question: Isn't Q24 enough for most cases?
> 
> The goal I had was for a Q16 version, since, I beliieve, that is the level most distros have. The MSYS2/mingw64 version was Q8 last time I checked ... if that were changed to Q16 the world would be a better place.

Yes I meant Q16. Then I will make another PR to build Q16 GraphicsMagick in the CI script.

maynardGK commented:
> 
> I don't think the difference in the XDR is a big problem. HDF4 just has its own implementation of XDR, which is dependent on other internal HDF4 headers, and this can be considered as something independent from the normal XDR library that we use (i.e. BSD-XDR). Also, the header location is different (one is in `include`, the other one is in `include/rpc`) and they do not overwrite each other.
> 

OK but  we need to note that lib/libxdr.dll.a is named the same for both, so it may be ok if used by bsd-xdr but probably a few more entry points when used for libxdr,dll (.vs. mgwxdr-0.dll).

GillesDuvert committed:
removed old slow and nonportable findfile (still used in many codes), replaced by a suitable subset of C++ file_search. Huge speed improvements expected on existing 'proven' ^)) historical codes.

GillesDuvert committed:
Merge origin/WidgetsVersion2 into WidgetsVersion2

GillesDuvert committed:
gdl: default is to use system fonts, added --widget-compat for compatiblity size
gdlwidget: widget_info (/geom) shoudl be ok now. Some improvements elsewhre.
devicewx: device queries (abused in coyote procedures) takes too much time if we use the current scheme. Lots of speedup possible here.

pjb7687 commented:
On Windows, the name of the plotting device is still 'X' and it seems like this needs to be changed to 'WIN'.
I have a question - after we merge this, libwxwidgets becomes one of the mandatory packages since this will replace the current devicex or devicewin, is this correct?

GillesDuvert commented:
> On Windows, the name of the plotting device is still 'X' and it seems like this needs to be changed to 'WIN'.
> I have a question - after we merge this, libwxwidgets becomes one of the mandatory packages since this will replace the current devicex or devicewin, is this correct?

Yes thanks for the opportunity to expose my plans:

1. I suggest to have wxWidgets a 'main' dependency. Removing the need of X11 for unix/linux/MacOS, as well as xwin and wingcc/wingdi . This by default. Behavior of plots, either in a WIDGET_DRAW or in a WINDOW, will be identical.
2. provided wxWidgets library exist, the (only) device thus existing will be 'X'. (can be if you prefer 'MAC' on macs and 'WIN' on windows, buts passing as 'X' is the safest solution IMHO since the behaviour of GDL is unique?). 
3. if the wxWidgets library does not exist, then it is a compilation failure unless cmake enforces X11 (-DX11=ON) (or -DWINGCC or equivalent for windows) . We get then the 'old' behaviour: no widgets, plot device is 'X' or 'WIN'.
4. If wxWidgets is present, and cmake enforces X11 (-DX11=ON)[or -DWINGCC=ON) , then by default behaviour is as in 2. unless the option `--no-use-wx` is used in calling gdl. With `--no-use-wx` the widgets will use wx, the plot window will use wxin or wingdi, the device will be 'X' or 'WIN' respectively, and the wxwidgets plot device will be available as 'MAC' (as of today).

GillesDuvert committed:
changed 'Debug' to 'Release' for speed tests

pjb7687 commented:
> > On Windows, the name of the plotting device is still 'X' and it seems like this needs to be changed to 'WIN'.
> > I have a question - after we merge this, libwxwidgets becomes one of the mandatory packages since this will replace the current devicex or devicewin, is this correct?
> 
> Yes thanks for the opportunity to expose my plans:
> 
> 1. I suggest to have wxWidgets a 'main' dependency. Removing the need of X11 for unix/linux/MacOS, as well as xwin and wingcc/wingdi . This by default. Behavior of plots, either in a WIDGET_DRAW or in a WINDOW, will be identical.
> 2. provided wxWidgets library exist, the (only) device thus existing will be 'X'. (can be if you prefer 'MAC' on macs and 'WIN' on windows, buts passing as 'X' is the safest solution IMHO since the behaviour of GDL is unique?).
> 3. if the wxWidgets library does not exist, then it is a compilation failure unless cmake enforces X11 (-DX11=ON) (or -DWINGCC or equivalent for windows) . We get then the 'old' behaviour: no widgets, plot device is 'X' or 'WIN'.
> 4. If wxWidgets is present, and cmake enforces X11 (-DX11=ON)[or -DWINGCC=ON) , then by default behaviour is as in 2. unless the option `--no-use-wx` is used in calling gdl. With `--no-use-wx` the widgets will use wx, the plot window will use wxin or wingdi, the device will be 'X' or 'WIN' respectively, and the wxwidgets plot device will be available as 'MAC' (as of today).

Thanks! I would say, additionally we might just want to deprecate wingcc/wingdi devices on Windows. Actually, these devices are too old, for example, they do not inherit GraphicsMultiDevice class and produce weird errors e.g. #841.

Regarding the name of the device, using 'WIN' as driver name on Windows is just IDL behavior (https://www.l3harrisgeospatial.com/docs/supported_devices.html). For this, we can just rename the new driver name on Windows using ifdef.

On Linux/macOS it is up to others' opinions, but in principle, I think it is better to deprecate X device too and focus on improving the wxWidgets-based codebase. Because in this case we only have one multi-platform single device for plotting/widgets, and this will be much easier to manage.

GillesDuvert commented:
close,reopen

GillesDuvert committed:
prevent ALL widgets to get focus at creation.
I believe it is better than the reverse. Observe behaviour of 'atv' (atv.pro in idlastro).

GillesDuvert committed:
pushed old devices in a separate directory
made compilation such as will use only wxWidgets for plots unless told differently
Most probably will not compile on some platforms

GillesDuvert committed:
Merge origin/WidgetsVersion2 into WidgetsVersion2

GillesDuvert committed:
oops.

GillesDuvert committed:
insure wxwidgets driver of plplot is found when compiling ---> will force distribution to be OK

GillesDuvert committed:
removed unused variables

GillesDuvert committed:
removed gcc warning

GillesDuvert committed:
simplified code

GillesDuvert committed:
removed oodles of unused variables

GillesDuvert committed:
removed oodles of unused variables

GillesDuvert committed:
enable --notebook (GDL uses SVG only) and permit a GDL without xwin, wxWidgets and windgi drivers to run

GillesDuvert committed:
oops

GillesDuvert commented:
work in progress: works definitely on linux & windows (slow on windows) . Severe problem on Mac OSX (GDL stuck in eventloop). Probably signals a (not unexpected, this!)  bad eventloop construction/use in GDL, solving it may certainly improve GDL efficiency.

GillesDuvert committed:
removed complications due to 'our' wxApp . should have done that since beginning.
this code seem to work on MSW and OSX as well

GillesDuvert committed:
finally OK on all 3 platforms. Uses ToolBar instead of menuBar (same functionalities on 3 platforms).
Subtle wxApp initalisation & loop behaviour are implemented for windows. Otherwise no need ow a wxApp apparently.

GillesDuvert commented:
normally this version has the same behaviour on OSX, Windows and linux, does not need X11 or wingcc, is fast etc.
Provided the installed version of plplot has the wxwidgets driver, everything should be ok.
If problems, will insert the wxwidgets plplot driver in gdl's code and we get rid of all thise problems once and 4 all.

GillesDuvert committed:
removed zenity oldies.
changes wmenu to use widgets
solved size problems (and geometry problems) on OSX.
Loop event solution for Windows terribly slow. to be solved.
linux and OSX ok
added script to recompile plplot with wxWidgets on OSX using brew.
Current cmake forces compilation to stop if wxwidgets driver is not found in plplot.

GillesDuvert committed:
bunch of simple obsolete commands (found in many old instrument pipelines) that can easily be transcribed to new commands.

GillesDuvert committed:
rewrote to express the priority for wxWidgets and still permit reconstuction on special cases. See internal comments.

GillesDuvert committed:
X11 is not set per default now.

GillesDuvert committed:
try to get wxwidgets driver in plplot, at least under unix!

GillesDuvert committed:
try again

GillesDuvert committed:
avoid strange cmake messages?

GillesDuvert committed:
try again

GillesDuvert commented:
OK, passes everywhere except for some tests that are plainly wrong and the 'mini' dep that evidently fails because the compilation conditions have changed (MINI must now be compiled with -DX11=ON or -DINTERACTIVE_GRAPHICS=OFF).
And, python tests fail for some silly reason i do not care at all.
I suggest to remove all the builds except the full macsox and the full linux + codecov, this will avoid many headaches. All other builds have no interest at all, the one and only GDL is with all options and one unified graphics through wxwidgets.

GillesDuvert committed:
looked into some python problems.

GillesDuvert committed:
add DINTERACTIVE_GRAPHICS=OFF for 'MINI' rebuilds

GillesDuvert committed:
oops the hell with the complicated cmake syntax...

alaingdl commented:
HI. I am not sure I fully understand you. 
What will happen on computing nodes on HPC where we have no wx and no X11 ?

GillesDuvert commented:
> HI. I am not sure I fully understand you.
> What will happen on computing nodes on HPC where we have no wx and no X11 ?

provided -DINTERACTIVE_GRAPHICS=OFF is present in the cmake command, everything will be fine.

GillesDuvert committed:
trick to have the PS and SVG outputs always (?) correctly closed, thus useable, whatever the current status of plot device.

GillesDuvert committed:
typo was fatal to clever brew trick!

GillesDuvert committed:
modified or sorted test procedures. Some are modified to pass on all platforms, many that uses (wrongly) unix commands in spawn are in spawn_is_dangerous (and should be rewritten such as avoiding os-dependent commands), some that call graphic devices commands (window, wset, wdelete) and will never work in MINI builds are set in interactive_tests/ and shoudl be run manually. Very wrong test procedures are in wrong/

GillesDuvert committed:
special test for osx

GillesDuvert committed:
revert, I'll never understand anything to these scripts.

GillesDuvert committed:
removed test_random, which is interactive.

GillesDuvert committed:
corrected small bug preventing to have INTS instead of DOUBLEs when some previous colum had blank values.

GillesDuvert committed:
raise plot panel after creation (no focus)

GillesDuvert committed:
erase working (wxwidgets), channel erase TO BE DONE.

GillesDuvert committed:
correct margins for wxwidgetw plots

GillesDuvert committed:
added warning

GillesDuvert committed:
solved a few unicode font problems, including the angstrom bug #862.
Now TTF fonts in wx plots are ok.

maynardGK reviewed (commented):
I can only build, not run, with the mingw/msys2 gcc 10 on win10: runtime doesn't find
the $PATH where shared libraries are kept.
I can download and run the github builds (shareds gotten from directory holding gdl.exe).
windows graphics with --use-wx still gives the error pop-up box with message
````
 ../wxWidgets-3.5.01/src/msw/dc.cpp(550) assert "hrgn region" failed
 in SetClippingHrgn(); invalid clipping region
````

GillesDuvert committed:
try have wxwidgets installed BEFORE plplot...

GillesDuvert committed:
say if x11 is present

GillesDuvert committed:
try be more, eh, precise?

GillesDuvert committed:
Merge branch 'master' into WidgetsVersion2

GillesDuvert committed:
try to enable at least a compilation on OSX. 

Brew is not coming with wxwidgets, ./scripts/brew_enable_wxwidgets should recompile it OK (I've done it on a mac) but that does not work on Travis for an obscure reason. I quit.

GillesDuvert commented:
(post commit) works:
![image](https://user-images.githubusercontent.com/33936193/111611369-abcfa880-87dc-11eb-979d-1dcc27eed3a1.png)
